containers/src/Data/Sequence/Internal.hs

```diff
@@ -650,13 +650,42 @@ type Digit23 a = Node a
 -- If we used a higher-order nested type with MPTC, we could probably use a
 -- class, but as it is we have to build up @map23@ explicitly through the
 -- recursion.
+--
+-- Note: @aptyMiddle@ is polymorphic-recursive on @b@ and @c@, not @a@.
+--
+-- Invariants:
+--
+-- > 1. Viewing the various trees as the lists they represent:
+-- >
+-- >    toListFTN result
+-- >      =  (firstf <$> (toListThinN m ++ toListD sf)
+-- >      ++ (map23  <$> toListFTE fs <*> (toListD pr ++ toListThinN m ++ toListD sf))
+-- >      ++ (lastf  <$> (toListD pr ++ toListThinN m))
+-- >
+-- > 2. s = size m + size pr + size sf
+-- >
+-- > 3. size (firstf y) = size (lastf y) = size (map23 x y) = size y
+-- >      for any (x :: a) (y :: b)
+--
+-- Projecting invariant 1 on sizes, using 2 and 3 to simplify, we have the
+-- following corollary.
+-- It is weaker than invariant 1, but it may be easier to keep track of.
+--
+-- > 1a. size result = s * (size fs + 1) + size m
+--
+-- In invariant 1, the types of the auxiliary functions are as follows:
+--
+-- > toListFTE   :: FingerTree (Elem a) -> [a]
+-- > toListFTN   :: FingerTree (Node c) -> [c]
+-- > toListThinN :: Thin (Node b) -> [b]
+-- > toListD     :: Digit12 b -> [b]
 aptyMiddle
-  :: (b -> c)
-     -> (b -> c)
-     -> (a -> b -> c)
-     -> FingerTree (Elem a)
-     -> Rigid b
-     -> FingerTree (Node c)
+  :: (b -> c)              -- ^ @firstf@
+  -> (b -> c)              -- ^ @lastf@
+  -> (a -> b -> c)         -- ^ @map23@
+  -> FingerTree (Elem a)   -- ^ @fs@
```

Comment:
I don't really like the name `fs`, because while that's good for `<*>` it's a bit strange for `liftA2`; I kind of want to call that `xs`. But what are `firstf` and `lastf`? Not sure how to give those sensible names.

Reply:
The binary function `a -> b -> c` could be just `f`, then the two `(b -> c)` functions are `f` applied to the first and last element of the toplevel `xs` (maybe called `ffirst`, `flast`?), and `fs` is the middle part of `xs`, so renamed to `midxs`.

```
xs = first <: midxs :> last
ffirst = f first
flast = f last
```

Reply:
Yes, that sounds better!

Reply:
Is it really the same `f`? If so, can we just pass the first and last `x`? I don't remember how this goes.

Reply:
Good question. The difference is that the current version has better sharing. Otherwise instead of `fmap firstf` you would write `fmap (f first)`. In the case of an `f` that can share data when partially applied, that would cause that data to be recomputed at every recursive call. I can add a mention of this in the comments too.

Reply:
Do we also get that sharing in the middle?

Reply:
We really need a section of the documentation devoted to details of the performance of the `Applicative` instance.

Reply:
> Do we also get that sharing in the middle?

I think so, `map23 x` will expand to `fmap @Node (... (fmap @Node (f x)) ...)` and the `fmap`'s will duplicate the closure with the one `x`. (It might be good to actually check that `fmap . map23`, in the recursive call of `aptyMiddle`, doesn't duplicate `map23 x`.) `map23` will be applied to each `x` in the middle once.

The problem with `firstf` and `lastf` is that they must be used multiple times to build the digits on the sides.

Reply:
Linear types might be relevant to track this. I'm wildly speculating.

Reply:
Oof. Not going for linear types here! Ouch! I have been thinking it would be awfully nice to see if we can use Liquid Haskell in this module, mostly to prove totality of `zipWith` and `chunksOf` (where partial matches are *essential* for performance).